src/content/en/updates/posts/2016/03/access-usb-devices-on-the-web.markdown

```diff
@@ -0,0 +1,319 @@
+}
+
+void loop() {
+  // Nothing here for now.
+}
+{% endhighlight %}
+
+The third-party [WebUSB Arduino
+library](https://github.com/webusb/arduino/tree/gh-pages/library/WebUSB) used
+in the sample code above does basically two things:
+
+- The device acts as a WebUSB device enabling Chrome to read
+  the [landing page URL](https://wicg.github.io/webusb/#webusb-platform-capability-descriptor)
+  and the [list of origins](https://wicg.github.io/webusb/#get-allowed-origins) allowed to communicate with.
+- It exposes a WebUSB Serial API that you may use to override the default one.
+
+Let's focus back to the JavaScript code. Once we get the `device` picked by the
+user, `device.open` simply runs all platform-specific steps to start a session
+with the USB device. Then, we have to select an available USB Configuration
+with `device.selectConfiguration`. Remember that a Configuration specifies how
+the device is powered, its maximum power consumption and its number of
+interfaces. Talking about interfaces, we also need to request exclusive access
+with `device.claimInterface` since data can only be transferred to an interface
+or associated endpoints when the interface is claimed. Finally calling
+`device.controlTransferOut` is needed to set up the Arduino device with the
+appropriate commands to communicate through the WebUSB Serial API.
+
+From there, `device.transferIn` performs a Bulk Transfer onto the
+device to inform that the host is ready to receive bulk data. Then, the promise
+is fulfilled with a `result` object containing a
+[DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView)
+`data` that has to be parsed appropriately.
+
+For those who are familiar with USB, all of this should look pretty familiar.
+
+### I want moar
+
+The WebUSB API lets you interact with the all USB transfer/endpoint types:
+
+- CONTROL transfers, used to send or receive configuration or command
+  parameters to a USB device are handled with `controlTransferIn(setup,
+  length)` and `controlTransferOut(setup, data)`.
+- INTERRUPT transfers, used for small amount of time sensitive data are handled
+  with same methods as BULK transfers with `transferIn(endpointNumber, length)`
+  and `transferOut(endpointNumber, data)`.
+- ISOCHRONOUS transfers, used for streams of data, like video and sound are
+  handled with `isochronousTransferIn(endpointNumber, packetLengths)` and
+ `isochronousTransferOut(endpointNumber, data, packetLengths)`.
+- BULK transfers, used to transfer a large amount of non-time-sensitive data in
+  a reliable way are handled with `transferIn(endpointNumber, length)` and
+ `transferOut(endpointNumber, data)`.
+
+You may also want to have a look at Mike Tsao's [WebLight
+project](https://github.com/sowbug/weblight) which provides a ground-up example
+of building a USB-controlled LED device designed for the WebUSB API (not using
+an Arduino here). You'll find hardware, software, and firmware.
+
+## Tips
```

Comment:
I forgot to mention this in the README.md for the Arduino library but on Linux you need to add a [udev rule](https://www.freedesktop.org/software/systemd/man/udev.html) to allow Chrome to open the device. To allow members of the `plugdev` group to access Arduinos it would look like this:

`SUBSYSTEM=="usb", ATTR{idVendor}=="2341", MODE="0664", GROUP="plugdev"`